boojum-cuda/src/poseidon.rs

```diff
@@ -417,31 +420,40 @@ pub fn gather_rows(
 
 pub fn gather_merkle_paths(
     indexes: &DeviceSlice<u32>,
-    values: &(impl DeviceMatrixChunkImpl<BaseField> + ?Sized),
-    result: &mut (impl DeviceMatrixChunkMutImpl<BaseField> + ?Sized),
+    values: &DeviceSlice<GoldilocksField>,
+    results: &mut DeviceSlice<GoldilocksField>,
     stream: &CudaStream,
 ) -> CudaResult<()> {
-    assert_eq!(values.cols(), CAPACITY);
-    assert_eq!(result.cols(), CAPACITY);
+    assert_eq!(values.len() % CAPACITY, 0);
+    let values_count = values.len() / CAPACITY;
+    assert!(values_count.is_power_of_two());
+    let log_values_count = values_count.trailing_zeros();
+    assert_ne!(log_values_count, 0);
+    let log_leaves_count = log_values_count - 1;
+    assert_eq!(results.len() % CAPACITY, 0);
+    let results_count = results.len() / CAPACITY;
     let indexes_len = indexes.len();
-    let values_rows = values.rows();
-    let result_rows = result.rows();
-    assert_eq!(result_rows % indexes_len, 0);
-    let layers_count = result_rows / indexes_len;
-    assert_eq!(values_rows, 1 << (layers_count + 1));
+    assert_eq!(results_count % indexes_len, 0);
+    let layers_count = results_count / indexes_len;
```

Comment:
minor suggestion: if we force caller to supply `layers_count` (or `merkle_cap_size`) as an extra argument, we can fully enforce `values` and `results`'s sizes conform to caller's expectations, instead of reverse engineering `layers_count` ourselves.

Reply:
added `layers_count` argument